### ISSUE TITLE
[Qwen3.5 GDN]: Blackhole L1 planning failures in wide depthwise Conv1D and GatedDelta prefill

### ISSUE THREAD

**Issue by Eunho-J** [tenstorrent/tt-metal#49793]
## Current status (2026-07-22)

This issue tracks two independent Blackhole L1-planning bugs exposed by the same Qwen3.5-9B GDN prefill workflow.

| Failure | Root cause | Fix |
| --- | --- | --- |
| Wide biasless depthwise Conv1D | HEIGHT_SHARDED keeps all 8,192 channels in each participating core and exceeds per-core L1 | #49694 |
| GatedDelta multi-token sequence scan | Caller-owned L1 tensors overlap the operation static circular-buffer region | #49768 |

- [ ] #49694 is ready for review at `9a30fed4f14`. The branch is being rebased and validated against current main before the updated history is pushed.
- [ ] #49768 is ready for review at `3847becb8a5`. It also requires a final rebase and maintainer-approved CI.
- The broader fold, generic Conv2D planner, DRAM-slicing, and ConvTranspose parity work is intentionally excluded from #49694 and remains in the stacked draft [Eunho-J/tt-metal#1](https://github.com/Eunho-J/tt-metal/pull/1).
- The combined integration branch is a historical validation artifact and must be refreshed after both component PRs are rebased.

Related model-enablement issue: #42828.

## Environment

- Hardware: Tenstorrent Blackhole P150
- Model: `Qwen/Qwen3.5-9B`
- Batch size: 1
- Mesh: 1x1
- Device configuration: `l1_small_size=24576`, 2 command queues
- Depthwise clean reference: `45d9b04261922699e130fadaecef368aa0a57ac9`
- GatedDelta clean reference: `05b5201cc4e3db0322fed8b3b1a0ee54f1c2b19b`

Both failures were reproduced against clean TT-Metal source and are not regressions from the combined custom checkout.

## Failure 1: wide biasless depthwise Conv1D

Biasless 1D depthwise convolutions select the specialized native implementation. The Qwen3.5-9B GDN shape is:

| Parameter | Value |
| --- | ---: |
| N | 1 |
| Input length | 7 |
| Output length | 4 |
| Channels | 8192 |
| Kernel | 4 |
| Groups | 8192 |
| Bias | None |

The original HEIGHT_SHARDED plan retains all channels on each participating core:

```text
Statically allocated circular buffers on core range [0-0 - 0-0]
grow to 1684352 B, which is beyond max L1 size of 1572864 B
```

The Qwen helper previously avoided the native path with a `D > 2048` FIR fallback. That workaround hid the invalid plan and made routing depend on a model-specific threshold.

#49694 uses the existing per-core L1 model to select BLOCK_SHARDED only when it reduces L1 use and actually distributes channels. Input, weights, and output preserve the same channel ownership; invalid layouts fail explicitly. No channel-count or architecture threshold is added, and the Qwen workaround is removed.

The shared native depthwise weight-plan contract also exposed a regression in the existing ConvTranspose2D weight-preparation path. The narrow compatibility fix is included in #49694; no new ConvTranspose2D functionality is added.

Focused reproduction:

```bash
pytest -v -s \
  tests/ttnn/unit_tests/operations/conv/test_conv1d.py::test_conv1d_depthwise_auto_shard_qwen35_9b
```

## Failure 2: GatedDelta sequence-scan circular-buffer collision

The original end-to-end record showed a 1,244-token prefill completing and a 2,546-token prefill failing after the first 2,048-token chunk:

```text
Prefill started (1244 tokens)
Prefill completed in 8.4s
Prefill started (2546 tokens)
Statically allocated circular buffers in program 868 clash with L1 buffers
on core range [0-0 - 3-1]. L1 buffer allocated at 1271808 and
static circular buffer region ends at 1360640
```

Failing path:

```text
Qwen36Model.prefill_paged
  -> prefill_layer_chunked(chunk_size=2048)
  -> chunk_gated_delta_rule_seq
  -> ttnn.transformer.gated_delta_attn_seq
  -> GatedDeltaAttnSeqDeviceOperation
  -> ProgramImpl::validate_circular_buffer_region
```

Clean-baseline boundary probes with the real 32-layer checkpoint:

| Total tokens | Tail after 2048 | Result |
| ---: | ---: | --- |
| 2048 | 0 | pass |
| 2049 | 1 | pass, recurrent tail |
| 2050 | 2 | fail, sequence-scan tail |
| 2560 | 512 | fail |
| 2561 | 513 | pass, DRAM selected by the old length heuristic |

At 2,050 tokens, the 22nd GDN fused-convolution state occupies `[1357824, 1363968)`. The sequence operation static circular-buffer region is `[111360, 1360640)`, producing a 2,816-byte overlap.

At 2,546 tokens, live `k`, `gate_raw`, `b_raw`, and `a_raw` tensors occupy the per-core range `[1271808, 1353728)`, which is also inside the static circular-buffer region.

The standalone sequence operation and the full-model first chunk both pass at 2,048 tokens. The collision is caused by caller-owned L1 tensors remaining live across the sequence launch, not by unsupported sequence length or the 2,048-token chunk size.

#49768 removes the sequence-length capacity heuristic. Multi-token sequence-scan inputs, normalization intermediates, and outputs use DRAM; single-token recurrent decode retains L1; other recurrent callers preserve input placement. The C++ kernel and prefill chunk size are unchanged.

Focused regression suite:

```bash
HF_MODEL=/path/to/Qwen3.5-9B pytest --timeout=900 -v -s \
  models/demos/blackhole/qwen36/tests/test_prefill.py \
  -k "len300_b512 or len400_b512 or len512_b512 or chunk1_tail289_b512 or chunk1_tail498_b512"
```

### Depthwise fix (#49694)

- Release build: passed.
- Changed-file pre-commit hooks: passed.
- Full Conv1D suite: 63 passed, 1 opt-in slow case skipped.
- Opt-in long-sequence depthwise route: passed.
- Four block-sharded regression variants: passed.
- Full ConvTranspose2D unit suite: 204 passed.
- ConvTranspose2D nightly, sweep, and ULP suites: 136 passed.
- MobileNetV2 pretrained PCC regression: passed.
- Real-checkpoint Qwen3.5 GDN suite: 5 passed; DeltaNet PCC 0.998904.
- Qwen2.5-0.5B 128-token paged prefill: passed; PCC 0.9999700434.

### GatedDelta fix (#49768)

- Release and TTNN-test build: passed.
- GDN contract/reference tests: passed.
- Eager and traced decode references: passed.
- Real Qwen3.5-9B 32-layer 2,546-token prefill: completed with finite logits and no L1 collision.
- Full prefill file: 20 passed, 1 expected long-context skip, and one unchanged len2000/b2048 argmax assertion failure.
- Clean upstream reproduces the same len2000/b2048 PCC 0.999919 and token pair 12/11, identifying that assertion as baseline numerical drift rather than a #49768 regression.
- 2,048-token median latency: 1,362.691 ms baseline versus 1,363.196 ms fixed (+0.037%).

### Combined validation

Before the latest main advances, the combined checkout passed a fresh Release build, six GDN tests, the Qwen-shaped `C=groups=8192`, `K=4` Conv1D case, and the real-checkpoint 32-layer 2,546-token prefill with finite logits and no L1 collision. The combined branch will be refreshed after both PR rebases.

## Remaining work

- Finish and push the current-main rebase of #49694 after hardware validation.
- Rebase #49768 onto current main and repeat its focused hardware validation.
- Run maintainer-approved upstream sanity/post-commit CI for both PRs.
- Refresh the combined integration branch from the two final PR heads.

## Expected

- Native wide depthwise Conv1D selects a modeled plan that fits per-core L1 without a model-specific fallback.
- Multi-token GatedDelta prefill does not place live caller tensors inside the sequence operation static circular-buffer region.
- Real Qwen3.5-9B 2,546-token prefill completes without an L1 collision.
- Unsupported layouts fail explicitly rather than silently switching algorithms.

## Discovery provenance

The wide depthwise Conv1D L1 overflow was identified and reported by Idea from MAGI. The GatedDelta sequence-scan collision was found independently during follow-up investigation.
